kernel/aster-nix/src/fs/utils/page_cache.rs

```diff
@@ -85,16 +87,204 @@ impl Debug for PageCache {
     }
 }
 
+struct ReadaheadWindow {
+    /// The window.
+    window: Range<usize>,
+    /// Look ahead position in the current window, where the readahead is triggered.
+    /// TODO: We set the `lookahead_index` to the start of the window for now.
+    /// This should be adjustable by the user.
+    lookahead_index: usize,
+}
+
+impl ReadaheadWindow {
+    pub fn new(window: Range<usize>) -> Self {
+        let lookahead_index = window.start;
+        Self {
+            window,
+            lookahead_index,
+        }
+    }
+
+    /// Gets the next readahead window.
+    /// Most of the time, we push the window forward and double its size.
+    ///
+    /// The `max_size` is the maximum size of the window.
+    /// The `max_page` is the total page number of the file, and the window should not
+    /// exceed the scope of the file.
+    pub fn next(&self, max_size: usize, max_page: usize) -> Self {
+        let new_start = self.window.end;
+        let cur_size = self.window.end - self.window.start;
+        let new_size = (cur_size * 2).min(max_size).min(max_page - new_start);
+        Self {
+            window: new_start..(new_start + new_size),
+            lookahead_index: new_start,
+        }
+    }
+
+    pub fn lookahead_index(&self) -> usize {
+        self.lookahead_index
+    }
+
+    pub fn readahead_index(&self) -> usize {
+        self.window.end
+    }
+
+    pub fn readahead_range(&self) -> Range<usize> {
+        self.window.clone()
+    }
+}
+
+struct ReadaheadState {
+    /// Current readahead window.
+    ra_window: Option<ReadaheadWindow>,
+    /// Maximum window size.
+    max_size: usize,
+    /// The last page visited, used to determine sequential I/O.
+    prev_page: Option<usize>,
+    /// Readahead requests waiter.
+    waiter: BioWaiter,
+}
+
+impl ReadaheadState {
+    const INIT_WINDOW_SIZE: usize = 4;
+    const DEFAULT_MAX_SIZE: usize = 32;
+
+    pub fn new() -> Self {
+        Self {
+            ra_window: None,
+            max_size: Self::DEFAULT_MAX_SIZE,
+            prev_page: None,
+            waiter: BioWaiter::new(),
+        }
+    }
+
+    /// Sets the maximum readahead window size.
+    pub fn set_max_window_size(&mut self, size: usize) {
+        self.max_size = size;
+    }
+
+    fn is_sequential(&self, idx: usize) -> bool {
+        if let Some(prev) = self.prev_page {
+            idx == prev || idx == prev + 1
+        } else {
+            false
+        }
+    }
+
+    /// The number of bio requests in waiter.
+    /// This number will be zero if there are no previous readahead.
+    pub fn request_number(&self) -> usize {
+        self.waiter.nreqs()
+    }
+
+    /// Checks for the previous readahead.
+    /// Returns true if the previous readahead has been completed.
+    pub fn check_for_prev_readahead(&self) -> bool {
```

Comment:
`check_for_prev_readahead` is not a good name for a function to return a bool value. 

I think `prev_readahead_is_completed` is better.

Reply:
Fixed.